### PR TITLE
HOSTEDCP-1773: Feature gates openshift markers

### PR DESCRIPTION
**Pull request by enxebre**
# Feature Gates

Feature gates in OpenShift allows to ensure everything knew works together and optimizes for rapid evaluation in CI to promote without ever releasing a public preview.
There are no guarantees that your fleet continues to be operational long term after you enable a feature gate.

In the HCP context, there are multiple non exclusive scenarios where features might need to be gated:

1 - A feature that impacts HO install
E.g. New CRDs are required for CPOv2 (control plane operator version 2)

2 - A feature that impacts the whole cluster fleet / HO
E.g. Introduce fleet wise shared ingress to be validated in targeted environments

3 - A feature that impacts individual clusters
E.g. Introduce using CPOv2 for some HC to develop feedback

4 - A feature that impacts API
E.g. Introduce a new provider like Openstack
E.g. Introduce a new field/feature like AWS tenancy

5 - A feature specific for an OCP component
Components honour existing standalone in-cluster OCP feature gate mechanisim

The currently ongoing implementation of feature gates relies on "k8s.io/component-base/featuregate" to enable devs to declare [granular gates for their features](https://github.com/openshift/hypershift/blob/9f5ccaef47cdcf9d2df91134571f1783e99e30fe/hypershift-operator/featuregate/feature.go).

All the feature gates are grouped in a single TechPreviewNoUpgrade feature set. Current implementation exposes this --tech-preview-no-upgrade flag in the CLI at install time

```
hypershift install --help
```
Will show among other flags:
```
--tech-preview-no-upgrade                        If true, the HyperShift operator runs with TechPreviewNoUpgrade features enabled
```

In a follow up we'll consider to introduce support to also signal --tech-preview-no-upgrade at the HC level.
Eventually support for at least 1, 2, 3 and 4 afromentioned scenarios will most likely converge into a single API.

This implementation aligns with standalone for API/CRDs generation via openshift/api codegen based on openshift markers

### CODE REVIEW

**Comment by csrwng** [comment]
> In a follow up we'll consider to introduce support to also signal --tech-preview-no-upgrade at the HC level

Don't we already have this with `.spec.configuration.featureGate` ?

**Comment by enxebre** [comment]
>Don't we already have this with .spec.configuration.featureGate?

For OCP featuresSet yes. For things use cases similar to 3 that impact the service operability rather than the guest cluster UX is not fully clear to me we want to couple with that.

File: hypershift-operator/featuregate/feature.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package featuregate
+
+import (
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/component-base/featuregate"
+)
+
+const (
+	// OpenStack is a feature gate for running clusters on OpenStack.
+	// owner: @username
```
Comment (bryan-cox) [inline hypershift-operator/featuregate/feature.go:10]: Could Emilien and/or Matthew be added here? Or would we want to add a team alias here?

**Review by csrwng — COMMENTED** [review]
Some comments. One thing I think we're missing is a contributor-level doc describing how we would go about adding a new feature that is gated.

File: api/hypershift/v1beta1/featuregates/featureGate-Hypershift-Default.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{
```
[the remark this answers is not in the document]
Comment (enxebre) [reply api/hypershift/v1beta1/featuregates/featureGate-Hypershift-Default.yaml:1]: I think we could. I can try to include here once I see e2e go green or follow up. Only reason is I just took small steps towards plumbing the o/api codegen stuff bottom up. Our use case (any case beyond o/api) is not necessarily fully supported and it'll be contributing driven.

File: api/hypershift/register.go [hunk 3]
```diff
@@ -1,3 +0,0 @@
-package hypershift
-
-const GroupName = "hypershift.openshift.io"
```
[the remark this answers is not in the document]
Comment (enxebre) [reply api/hypershift/register.go:3]: The string needed to be inlined in the API for codegen to be able to parse it. You're right I can keep it for vendoring consumers.

File: api/hypershift/v1beta1/featuregates/featureGate-SelfManagedHA-Default.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+{
```
[the remark this answers is not in the document]
Comment (enxebre) [reply api/hypershift/v1beta1/featuregates/featureGate-SelfManagedHA-Default.yaml:1]: https://github.com/openshift/api/issues/2065

File: cmd/install/install.go [hunk 5]
```diff
@@ -563,11 +565,30 @@ func setupCRDs(opts Options, operatorNamespace *corev1.Namespace, operatorServic
...
 		crds, assets.CustomResourceDefinitions(
-			func(path string) bool {
+			func(path string, crd *apiextensionsv1.CustomResourceDefinition) bool {
 				if strings.Contains(path, "etcd") && opts.ExcludeEtcdManifests {
 					return false
 				}
+
+				// If the feature generated CRD has any featureSet version then it has the format nodepool-<featureSet>.
+				if strings.Contains(path, "zz_generated.crd-manifests") {
+					if pathSlice := strings.Split(path, "-"); len(pathSlice) > 1 {
```
[the remark this answers is not in the document]
Comment (enxebre) [reply cmd/install/install.go:575]: hah, thanks! that was preventing some CRDs from being installed. I fixed it and covered with additional test cases.

**Comment by enxebre** [comment]
>Some comments. One thing I think we're missing is a contributor-level doc describing how we would go about adding a new feature that is gated.

The openstack use case is fairly trivial since most of it happens at crd install time. However there might be features that require conditional business logic in some/all of our controllers. I covered this in the original autoprovisioning PR
https://github.com/openshift/hypershift/pull/4411/commits/bc85aef1f16b5d759799b96a2751a70372f8d545#diff-0b74c3eef0cb0197fb13878e7c543fc612b93ce47369c220668158ffd39a072fR7-R15
https://github.com/openshift/hypershift/pull/4411/commits/d3f4635d3cd6b263ecb2ce5f0514b48cb5d5e9b1
The idea was to accommodate feature gated behaviours into isolated files/controllers as much as possible so they were easily identifiable. 
I'm also planning to follow up to isolated the OpenStack API and maybe their logic bits. I'd prefer we wait until we implement that or even we introduce a more sophisticated use like autoprovisioning to flesh out a dev guide.

File: hypershift-operator/featuregate/feature.go (hunk 1 above)
Comment (enxebre) [reply hypershift-operator/featuregate/feature.go:10]: let me follow up to introduce team alias.

**Pull request merged**